### ISSUE TITLE
SC2096 complains about multiple arguments in my shebang, but it always works on my box

### ISSUE THREAD

**Issue by hborders** [koalaman/shellcheck#833]
#### For bugs
- Rule Id (if any, e.g. SC1000):  SC2096
- My shellcheck version (`shellcheck --version` or "online"): "online"
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash -euo pipefail
```

#### Here's what shellcheck currently says:

```
Line 1:
#!/bin/bash -euo pipefail
^-- SC2096: On most OS, shebangs can only specify a single parameter.
```

#### Here's what I wanted or expected to see:

I didn't expect any warning here. Multiple arguments work fine in bash on OS X 10.12.3 and the `GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin16)` it ships with.

**Comment by kurahaupo** [comment]
That's fairly clearly a feature of OS X; it definitely _won't_ work on most other operating systems.

Use
`set -u -o pipefail`

Just avoid `-e` entirely; it has awful corner cases where it doesn't fail where it should, and does fail where it shouldn't, and ignores the fact that non-zero status may simply be that: a status, rather than an exception; you're much better off actually checking the status of any important commands.

And `-u` is redundant after you've run shellcheck -- it should spot any misspelt variables for you.

**Comment by hborders** [comment]
Should there be an option to only warn about issues that affect your current OS? This script will only run on OS X.

**Comment by koalaman** [comment]
Huh, I didn't know OSX allowed that. I thought I had tested it.

There is a small handful of checks that would be better if ShellCheck was aware of the OS the script targets, and this is now one of them. Optimally you should be able to specify `#shellcheck os=macos` or similar.

**Comment by nihilus** [comment]
Actually this behaviour is not OS X specific, it is rather up to the interpreter to handle or not handle arguments in most OSes in general and Linux in particular. The normal way to use haserl is to give some arguments to limit file uploads and such.

**Comment by lilyball** [comment]
I run into this one every now and then and would love a way to instruct shellcheck globally to assume macOS as nearly all shell scripts I write are intended for macOS specifically.

**Comment by kurahaupo** [comment]
@nihilus The reason for the "non-failure" _is_ OS-specific.
Interpreters - like all programs - rightly expect separate arguments as separate elements in argv. On _most_ OSes, the interpreter gets given the text after the interpreter name as just one element in argv, with embedded spaces. _That_ is why it fails as I showed above.

Any suggestion that the interpreter must parse the text string is deeply flawed.

In my opinion this is an over-use of the shbang to do something that should be controlled using the `set` built-in.

**Comment by mnpenner** [comment]
What about

```
#!/usr/bin/env -S bash -fuo pipefail
```

The `-S` is specifically documented as to split the args:

```
       -S, --split-string=S
              process and split S into separate arguments; used to pass multiple arguments on shebang lines
```

But I don't know if that flag is supported by `env` on all systems either. It's working fine on Ubuntu.

**Comment by koalaman** [comment]
Since this bug was originally filed, ShellCheck has started supporting `env -S` (which is also supported on macOS), and you can state that all your scripts will only ever run on macOS with `echo disable=SC2096 >> ~/.shellcheckrc`.
